rclcpp/include/rclcpp/intra_process_manager_impl.hpp

```diff
@@ -184,6 +186,7 @@ class IntraProcessManagerState : public IntraProcessManagerStateBase
     size_t & size
   )
   {
+    std::lock_guard<std::mutex> lock(runtime_mutex_);
```

Comment:
To implement this more optimally, I could have one mutex for `publishers_` and then a mutex for each `PublisherInfo` entry in `publishers_`. That way one thread could look up an entry in `publishers_` while the other is looking up something in the map owned by an entry in `publishers` (I believe that would be fine). And yes, that implementation would include moving the mutex for `publishers_` into the block where `find` is invoked one the map.